Handle paths with/out slash prefix interchangeably.

jankeromnes opened the pull request:
This change allows the following alternatives:

    server.path('hello', function (req, res) { res.end('hi'); });

and

    server.path('/hello', function (req, res) { res.end('hi'); });

which are equivalent.

This is to keep `camp` compatible with [Self API](https://github.com/jankeromnes/selfapi). @espadrine please take a look.

bnjbvr commented on lib/camp.js:
```diff
@@ -712,6 +712,10 @@ function Path(matcher) {
   } else if (typeof matcher === 'string') {
+    if (matcher[0] !== '/') {
+      matcher = '/' + matcher;
+      this.matcher = matcher;
```
Probably cleaner to define `this.matcher` once and for all (remove the `this.matcher =` at the top) at the end of this function's body instead?

espadrine replied:
The idea is to have `this.matcher` be as close to the source material as possible, to allow introspection (which handler produces this?) and planned performance improvements.

jankeromnes replied:
Ok, I don't need to change it then.

jankeromnes commented:
Thank you @bnjbvr for the drive-by review! I'll just wait for @espadrine's input on whether we should keep or remove `this.matcher`.

espadrine commented:
`GET // HTTP/1.1` is a [valid](https://tools.ietf.org/html/rfc2396) HTTP Request-Line, and I wish to support it. This patch breaks support for it.

In fact, in retrospect, not requiring a slash at the start of string paths may have been a blunder that I intend to fix. Indeed, first, it hides the fact that this is an absolute URI; second, `OPTIONS * HTTP/1.1` is also valid, and does not start with a slash.

On the plus side, the end of the year is near, so version 17 (a semver breaking change) seems OK. To limit breakage, we can insert a slash when the path does not start with a slash or a star.

What do you think?

jankeromnes commented:
> Probably cleaner to define this.matcher once and for all (remove the this.matcher = at the top) at the end of this function's body instead?

Changed my code to not change `this.matcher` at all (so that it stays just like the constructor parameter for future introspection).

> In fact, in retrospect, not requiring a slash at the start of string paths may have been a blunder that I wish to fix.

I agree that it makes more sense to have everyone specify the first slash. Even without the caveats, it makes absolute URLs easier to understand.

> To limit breakage, we can insert a slash when the path does not start with a slash or a star.
> 
> What do you think?

Agreed, added a check for `matched[0] === '*'`. Are there any other special chars like this? What about `\`?

jankeromnes commented:
@espadrine PTAL

jankeromnes commented:
(rebased)

jankeromnes commented:
Now depends on #56.
